Add C++17 filesystem compatibility header

msimberg opened the pull request:
Adds a basic compatibility layer for C++17 filesystem. Mostly just forwards `std::filesystem` to `hpx::filesystem` if it's available and otherwise forwards `boost::filesystem` instead. The appropriate `error_code` is introduced into the `hpx::filesystem` namespace.

~It's not 100% transparent because of `error_code`. I've so far put explicit ifdefs where needed. The other main difference is that `canonical` doesn't take a base path for relative paths. In some cases we set this to `initial_path` but I'm not sure we care about the difference between that and `current_path`. I could probably put together something that does the same with `std` features.~

Where no naming changes were required I've just assumed that the functions behave the same, but let's see first if the tests even pass.

If someone wants to try `std::filesystem` it requires GCC 9, or an older GCC and explicit linking to `stdc++fs`. I think it's not yet in any released `libc++` but it can be linked to explicitly with `c++fs` as well.

Part of #3440.

hkaiser commented on src/runtime/components/server/runtime_support_server.cpp:
```diff
@@ -1581,13 +1580,12 @@ namespace hpx { namespace components { namespace server
                     fs::path lib_path =
                         lib / std::string(HPX_MAKE_DLL_STRING(component));
-                    if(fs::exists(lib_path, fsec))
+                    if(fs::exists(lib_path))
```
This will throw an exception, no? I think for the purpose of the filesystem API we can introduce a `filesystem::error_code` that maps onto the appropriate type (`std` or `boost`) and use that for the filesystem API calls we have.

hkaiser commented on src/runtime/components/server/runtime_support_server.cpp:
```diff
@@ -2015,13 +2013,12 @@ namespace hpx { namespace components { namespace server
                     fs::path lib_path =
                         lib / std::string(HPX_MAKE_DLL_STRING(component));
-                    if(fs::exists(lib_path, fsec))
+                    if(fs::exists(lib_path))
```
same as above

hkaiser commented on src/util/init_ini_data.cpp:
```diff
@@ -39,8 +36,12 @@ namespace hpx { namespace util
+            std::error_code ec;
+#else
             boost::system::error_code ec;
+#endif
```
Yah, this confirms it. Let's introduce `hpx::filesystem::error_code` instead.

msimberg replied:
Good idea. Thanks!

msimberg commented:
Updated. The error code now comes from either `std` or `boost` depending on which `filesystem` is used. I also added the wrappers for `canonical`, `initial_path` and `basename` that exist in `boost::filesystem` but not in `std::filesystem`. Boost.Filesystem is also not linked to if `filesystem` exists in the standard library.

hkaiser commented on libs/filesystem/include/hpx/filesystem.hpp:
```diff
@@ -0,0 +1,76 @@
+        return p.stem();
+    }
+
+    inline path canonical(path p, path base)
```
Couldn't these be passed by `const&` instead?

msimberg replied:
Yep, and should probably be passed like that as well. That's what the `std::filesystem::canonical` takes anyway. Good catch, thanks!

hkaiser commented on libs/filesystem/include/hpx/filesystem.hpp:
```diff
@@ -0,0 +1,76 @@
+        return ip;
+    }
+
+    inline path basename(path p)
```
Couldn't this be passed by `const&` instead?
